### ISSUE TITLE
Delayed typing in large file

### ISSUE THREAD

**Issue by muirdm** [zenspider/enhanced-ruby-mode#171]
The change in 3b97c6f7c4e0e462b74d57d3a0164f6b6f9b498e has resulted in terrible performance for me on large files. Typing anywhere in the large file is very delayed. It seems like the communication with the background process has to be asynchronous to not impact typing latency.

**Comment by zenspider** [comment]
I was just about to reference a previous performance issue.... but it was you. 😛

#146

**Comment by zenspider** [comment]
Can you do the profiling I did in #146 with your representative large file? 

I'm totally open to suggestion on how best to handle the situation that this change fixes, as long as it doesn't regress. I had all sorts of pain from errors on "dead" buffers and this ensured they were processed and done by the time I could kill them. Better error handling (I tried several different styles) or whatever...

**Comment by muirdm** [comment]
```
(with-current-buffer "big_file.rb"
  (profiler-start 'cpu)
  (--dotimes 10 (self-insert-command 1 ?s))
  (profiler-report)
  (profiler-stop))
```

With the `before-change-functions` call to `erm-wait-for-parse` (very slow):

```
- ...                                                             256 100%
 - progn                                                          256 100%
  - save-current-buffer                                           256 100%
   - let                                                          256 100%
    - unwind-protect                                              176  68%
     - progn                                                      176  68%
      - erm-parse                                                 176  68%
       - let                                                      176  68%
        - cond                                                    176  68%
         - if                                                     176  68%
          - erm-syntax-response                                   175  68%
           - save-excursion                                       175  68%
            - let                                                 175  68%
             - while                                              175  68%
              - let                                               175  68%
               - if                                               175  68%
                - progn                                           175  68%
                 - if                                             175  68%
                  - progn                                         156  60%
                   - save-current-buffer                          156  60%
                    - let                                         156  60%
                     - unwind-protect                             156  60%
                      - progn                                     156  60%
                       - erm-parse                                156  60%
                        - let                                     156  60%
                         - cond                                   156  60%
                          - setq                                  156  60%
                           - condition-case                       156  60%
                            - catch                               156  60%
                             - if                                 156  60%
                              - enh-ruby-add-faces                156  60%
                               - let*                             118  46%
                                - if                              118  46%
                                 - while                           61  23%
                                  - let                            51  19%
                                   - while                         49  19%
                                    - put-text-property                  8   3%
                                       car                          3   1%
                                    - setq                          3   1%
                                       cdr                          1   0%
                                 - if                              57  22%
                                  - progn                          57  22%
                                   - while                         39  15%
                                    - put-text-property                  4   1%
                                       1-                           1   0%
                                    - setq                          3   1%
                                       cdr                          1   0%
                                      remove-text-properties                  3   1%
                                 car                               38  14%
                  - move-end-of-line                               19   7%
                   - line-move                                     19   7%
                      line-move-1                                  19   7%
          - if                                                      1   0%
           - if                                                     1   0%
            - erm-do-syntax-check                                   1   0%
             - if                                                   1   0%
              - let                                                 1   0%
               - if                                                 1   0%
                - save-current-buffer                               1   0%
                 - if                                               1   0%
                    progn                                           1   0%
    - while                                                        80  31%
     - call-interactively                                          80  31%
      - debug                                                      80  31%
       - recursive-edit                                            80  31%
        - command-execute                                          80  31%
         - call-interactively                                      80  31%
          - funcall-interactively                                  80  31%
           - eval-last-sexp                                        80  31%
            - elisp--eval-last-sexp                                80  31%
             - eval                                                80  31%
              - progn                                              80  31%
               - save-current-buffer                               80  31%
                - let                                              79  30%
                 - while                                           79  30%
                  - self-insert-command                            79  30%
                   - #<lambda 0x23c0dc7cbe0b6d>                    79  30%
                    - erm-wait-for-parse                           79  30%
                     - while                                       79  30%
                      - accept-process-output                       5   1%
                       - erm-filter                                 5   1%
                          setq                                      5   1%
                  profiler-report                                   1   0%
   Automatic GC                                                     0   0%
```

Without the call (not slow):
```
- ...                                                               2 100%
 - progn                                                            2 100%
  - save-current-buffer                                             2 100%
   - let                                                            2 100%
    - while                                                         2 100%
     - call-interactively                                           2 100%
      - debug                                                       2 100%
       - recursive-edit                                             2 100%
        - command-execute                                           2 100%
         - call-interactively                                       2 100%
          - funcall-interactively                                   2 100%
           - eval-last-sexp                                         2 100%
            - elisp--eval-last-sexp                                 2 100%
             - eval                                                 2 100%
              - progn                                               2 100%
               - save-current-buffer                                2 100%
                - let                                               1  50%
                 - while                                            1  50%
                  - self-insert-command                             1  50%
                   - erm-req-parse                                  1  50%
                    - let                                           1  50%
                     - setq                                         1  50%
                      - catch                                       1  50%
                         if                                         1  50%
                  profiler-report                                   1  50%
   Automatic GC                                                     0   0%
```

**Comment by muirdm** [comment]
> I'm totally open to suggestion on how best to handle the situation that this change fixes, as long as it doesn't regress.

Do you have a way to reproduce the dead buffer errors reliably?

**Comment by zenspider** [comment]
Not a stable one... it’s mostly comes from rapidly modifying saving and closing like with macros across files.

**Comment by muirdm** [comment]
Do you have an example of an error?

**Comment by zenspider** [comment]
I don't. I'm disabling 3b97c6f locally for a while to see if I can replicate it again.

**Comment by zenspider** [comment]
Circling back (sorry)... Ironically, while doing more benchmarking I did get a consistent reproduction of the saving//closing error:

```elisp
(with-current-buffer "ruby25_parser.rb"
  (goto-char (point-max))

  ;; (electric-indent-mode (if electric-indent-mode -1 1))
  ;; (electric-indent-mode -1)
  ;; (electric-indent-mode 1)

  ;; (erm-wait-for-parse)
  ;; (message "starting")
  (profiler-start 'cpu)

  (--dotimes 10 (call-interactively 'newline))
  ;; (erm-wait-for-parse)

  (profiler-report)
  (profiler-stop)

  (with-current-buffer "ruby25_parser.rb"
    (revert-buffer nil t))
  )
```

with that `revert-buffer` it blows up every time. I'll try to take a whack at it at some point, but I'll deal with it since it is probably mostly hitting me.

**Comment by zenspider** [comment]
Also interesting: with that benchmark above I'm getting these 3 different results:

```elisp
;; electric-indent-mode -1 : 5 samples
;; electric-indent-mode  1 : 147, 139, 147 samples
;; NEITHER?!?!             : 193, 194, 150 samples ?!?
```

I dunno what's going on or why, but when I don't specify either electric-indent-mode, it seems slower than when I explicitly turn it on?!? Dunno.

**Comment by zenspider** [comment]
OK. I've rolled back the code that made it slow... I've profiled a fair amount including up to ~25kloc. If electric-mode is off, this all seems pretty zippy. Calling `erm-wait-for-parse` will slow things down, but seems to be dependent on where you are in the file (oddly, being at the end of the file seems faster? maybe it does a subset?)

Let me know if you see anything to the contrary. I have an OK setup now to profile before/after scenarios.
